contracts/StakeManagerV2.sol

```diff
@@ -0,0 +1,356 @@
+    ) external initializer {
+        if (_operatorRegistry == address(0)) revert ZeroAddress();
+        if (_bnbX == address(0)) revert ZeroAddress();
+
+        __AccessControl_init();
+        __Pausable_init();
+        __ReentrancyGuard_init();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        OPERATOR_REGISTRY = IOperatorRegistry(_operatorRegistry);
+        BNBX = IBnbX(_bnbX);
+    }
+
+    /// @notice Delegate BNB to the preferred operator.
+    /// @param _referralId referral id of KOL
+    /// @return The amount of BnbX minted.
+    function delegate(
+        string calldata _referralId
+    ) external payable override whenNotPaused nonReentrant returns (uint256) {
+        if (msg.value < STAKE_HUB.minDelegationBNBChange())
+            revert DelegationAmountTooSmall();
+
+        uint256 amountToMint = convertBnbToBnbX(msg.value);
+        BNBX.mint(msg.sender, amountToMint);
+
+        address preferredOperatorAddress = OPERATOR_REGISTRY
+            .preferredDepositOperator();
+        STAKE_HUB.delegate{value: msg.value}(preferredOperatorAddress, true);
+
+        emit Delegated(preferredOperatorAddress, msg.value);
+        emit DelegateReferral(msg.sender, msg.value, amountToMint, _referralId);
+        return amountToMint;
+    }
+
+    /// @notice Request to withdraw BnbX and get BNB back.
+    /// @param _amount The amount of BnbX to withdraw.
+    /// @return The index of the withdrawal request.
+    function requestWithdraw(
+        uint256 _amount
+    ) external override whenNotPaused nonReentrant returns (uint256) {
+        if (_amount == 0) revert ZeroAmount();
+
+        withdrawalRequests.push(
+            WithdrawalRequest({
+                user: msg.sender,
+                amountInBnbX: _amount,
+                claimed: false,
+                batchId: 0,
+                processed: false
+            })
+        );
+
+        uint256 requestId = withdrawalRequests.length - 1;
+        userRequests[msg.sender].push(requestId);
+
+        BNBX.transferFrom(msg.sender, address(this), _amount);
+        emit RequestedWithdrawal(msg.sender, _amount);
+
+        return requestId;
+    }
+
+    function startUndelegation(
+        uint256 _batchSize,
+        address _operator
+    )
+        external
+        override
+        whenNotPaused
+        nonReentrant
+        onlyRole(OPERATOR_ROLE)
+    {
+        if (_operator == address(0)) {
+            // if operator is not provided, use preferred operator
+            _operator = OPERATOR_REGISTRY.preferredWithdrawalOperator();
+        }
+        if (!OPERATOR_REGISTRY.operatorExists(_operator))
+            revert OperatorNotExisted();
+
+        address creditContract = STAKE_HUB.getValidatorCreditContract(
+            _operator
+        );
+        uint256 pooledBnb = IStakeCredit(creditContract).getPooledBNB(
+            address(this)
+        );
+
+        uint256 processedCount;
+        uint256 amountInBnbXToBurn;
```

Comment:
- move `lastIndex++`  inside.
- also why is checking necessary? `if (!withdrawalRequests[lastIndex].processed)`
```
while (lastIndex < withdrawalRequests.length && processedCount < _batchSize) {
    if (pooledBnb >= convertBnbXToBnb(amountInBnbXToBurn + withdrawalRequests[lastIndex].amountInBnbX)) {
        amountInBnbXToBurn += withdrawalRequests[lastIndex].amountInBnbX;
        withdrawalRequests[lastIndex].processed = true;
        withdrawalRequests[lastIndex].batchId = batchWithdrawalRequests.length;
        processedCount++;
        lastIndex++;
    } else {
        break;
    }
}
```